aiven/datasource_service_component.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2019 Aiven, Helsinki, Finland. https://aiven.io/
+package aiven
+
+import (
+	"fmt"
+	"github.com/aiven/aiven-go-client"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"log"
+)
+
+func datasourceServiceComponent() *schema.Resource {
+	return &schema.Resource{
+		Read: datasourceServiceComponentRead,
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Project name",
+			},
+			"service_name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Service name",
+			},
+			"component": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Service component name",
+			},
+			"route": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Network access route",
+			},
+			"host": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "DNS name for connecting to the service component",
+			},
+			"port": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Port number for connecting to the service component",
+			},
+			"kafka_authentication_method": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Kafka authentication method. This is a value specific to the 'kafka' service component",
+			},
+			"ssl": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Description: "Whether the endpoint is encrypted or accepts plaintext. By default endpoints are " +
+					"always encrypted and this property is only included for service components they may " +
+					"disable encryption",
+			},
+			"usage": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "DNS usage name",
+			},
+		},
+	}
+}
+
+func datasourceServiceComponentRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*aiven.Client)
+
+	projectName := d.Get("project").(string)
+	serviceName := d.Get("service_name").(string)
+	componentName := d.Get("component").(string)
+	route := d.Get("route").(string)
+
+	service, err := client.Services.Get(projectName, serviceName)
+	if err != nil {
+		return fmt.Errorf("service %s/%s not found: %w", projectName, serviceName, err)
+	}
+
+	for _, c := range service.Components {
+		if c.Component == componentName && c.Route == route && c.Usage == "primary" {
```

Comment:
Usage should be possible to define as data source property so that one can get MySQL/PostgreSQL/Redis master and standby info as well. Possible values are primary, syncing (node is still being initialized but already accessible) and replica.

Reply:
will add that